File: schemas/project-management/drizzle/task_dependencies.ts
// task_dependencies: directed relationships between tasks such as blocking or duplication.
// See README.md for full design rationale.

import { pgTable, pgEnum, uuid, timestamp, uniqueIndex, index } from "drizzle-orm/pg-core";
import { tasks } from "./tasks";

export const taskDependencyTypeEnum = pgEnum("task_dependency_type", [
  "blocks",
  "is_blocked_by",
  "relates_to",
  "duplicates",
]);

export const taskDependencies = pgTable(
  "task_dependencies",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    taskId: uuid("task_id").notNull().references(() => tasks.id, { onDelete: "cascade" }),
    dependsOnId: uuid("depends_on_id").notNull().references(() => tasks.id, { onDelete: "cascade" }),
    type: taskDependencyTypeEnum("type").notNull(),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    uniqueIndex("idx_task_dependencies_task_id_depends_on_id_type").on(table.taskId, table.dependsOnId, table.type),
    index("idx_task_dependencies_depends_on_id").on(table.dependsOnId),
  ]
);

File: schemas/project-management/spacetimedb/task_dependencies.rs
// task_dependencies: Dependency relationships between tasks.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum TaskDependencyType {
    Blocks, // type: String
    IsBlockedBy,
    RelatesTo,
    Duplicates,
}

#[spacetimedb::table(name = task_dependencies, public)]
pub struct TaskDependency {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub task_id: String, // UUID — FK → tasks.id (cascade delete)
    #[index(btree)]
    pub depends_on_id: String, // UUID — FK → tasks.id (cascade delete)
    pub dependency_type: TaskDependencyType,
    pub created_at: Timestamp,
    // Composite unique: (task_id, depends_on_id, type)
}
